### DISCUSSION TITLE
단위 테스트의 기준 -  단독(Solitary)인가 협동(Sociable)인가??

### DISCUSSION THREAD

**Discussion by sihyung92** [woowacourse-tobi-spring-study/tobi-vol1#50] [Show and tell]
마크의 2장 정리 가장~~ 하단에 있던 내용입니다!! [링크](https://github.com/woowacourse-tobi-spring-study/tobi-vol1/pull/40)

--- 

### 두번째. 단위테스트의 '단위'의 기준이 무엇인가??
마틴 파울러의 블로그 글을 바탕으로 정리해보았어요!
관심 있으신 분은 [여기](https://github.com/binghe819/TIL/blob/master/Test/TestGuide/%EB%8B%A8%EC%9C%84%ED%85%8C%EC%8A%A4%ED%8A%B8.md)를 참고해주세요!
그리고 각자 Solitary와 Sociable중 뭐를 더 선호하시는지를 이유와 같이 공유하면 좋을 것 같아요!

---
- 3줄 요약
1. 단위테스트는 사람마다 정의가 다르다. 그 중 `단위` 란 말의 정의는 상황적(팀바팀)이다. 단위를 객체 하나라고 정의 할 수도, 모듈 하나라고 정의할 수 도 있다. 
2. 그보다 더 중요한 건 단위를 철저히 격리(Mock 활용)하여 테스트하는지, 실체 프로덕션 객체를 이용하여 통합된 테스트를 활용하는지 이다.
3. 전자는 Solitary, 후자는 Sociable이라고 한다.

더 선호하는 테스트 방식에 대한 의견을 달아주심 되겠습니다~~

**Comment by sihyung92** [comment]
사실 Sociable해지는 순간 다니테스트가 아니지 않남?ㅎ_ㅎ 하고 생각하긴 하지만!

갠적으로 Solitary를 선호함~~ 빠른 피드백, 테스트 비용이 적음(금방 실행됨)
사실 Solitary가 Sociable보다 안 좋은 점은 격리시키기 어려운 구조인 경우(결합도가 높은 경우) 테스트 하기 어렵다.. 뿐이라고 생각하는데 Spring (DI 컨테이너)이나 Mockito같은 mock framework가 있는 상황에선 Solitary테스트를 만들기 좋기 때문에 Solitary로 가야한다고 봄~~

하지만 각 다니의 성공이 집합의 성공을 보장하진 않는다는 점 때문에! 다니테스트 위주로 구현하되 마지막 최종단에 통합 테스트는 필요하다고 생각함 
정말로 내가 안심하고 자려면 통합테스트가 성공해야 한다고 봄

**Reply by woowahanCU** [reply]
mockist와 classicist의 논쟁을 중점으로 봐도 좋겠네요. 
어려운 논의네요~ ㅎㅎ 전에 엘옵에서 잠시 논의된 적있어서~

> https://github.com/woowacourse-elegant-object/elegant-object-study/pull/86#discussion_r623618465

**Comment by Be-poz** [comment]
저도 solitary를 더 선호합니다. 일단 테스트 짜기가 더 편하고 딱 그 결과값만 보면 되기 때문에! sociable의 경우에는 말그대로 협동이기 때문에 신경쓸 것들이 많은게 상당히 귀찮은 것 같아요. service layer에 대한 테스트나 e2e 를 어쩔 수 없이 진행하게되면 그게 곧 sociable이긴 하지만요...  
그냥 둘 다 테스트 작성을 하는게 맘편한 것 같습니다!(실무에서 이렇게까지 신경쓸 수 있을까? 라는 불안함도 공존하네요 .. 🤔  )
